pennylane/pulse/transmon.py

```diff
@@ -215,3 +217,285 @@ def __add__(self, other):
+
+        if self.phase_is_callable:
+
+            def callable_phase(params, t):
+                return amp * trig_fn(phase(params, t) + freq * t)
+
+            self.func = callable_phase
+            return
+
+        if self.freq_is_callable:
+
+            def callable_freq(params, t):
+                return amp * trig_fn(phase + freq(params, t) * t)
+
+            self.func = callable_freq
+            return
+
+        # 0 out of 3 callable
+        # (the remaining coeff is still callable due to explicit time dependence)
+
+        def no_callable(_, t):
+            return amp * trig_fn(phase + freq * t)
+
+        self.func = no_callable
+
+    def __call__(self, params, t):
+        return self.func(params, t)
+
+
+def _reorder_AmpPhaseFreq(params, coeffs_parametrized):
+    """Takes `params`, and reorganizes it based on whether the Hamiltonian has
+    callable phase and/or callable amplitude and/or callable freq.
+
+    Consolidates amplitude, phase and freq parameters in they are callable,
+    and duplicates parameters since they will be passed to two operators in the Hamiltonian"""
+
+    reordered_params = []
+
+    coeff_idx = 0
+    params_idx = 0
+
+    for i, coeff in enumerate(coeffs_parametrized):
+        if i == coeff_idx:
+            if isinstance(coeff, AmplitudeAndPhaseAndFreq):
+                is_callables = [
+                    coeff.phase_is_callable,
+                    coeff.amp_is_callable,
+                    coeff.freq_is_callable,
+                ]
+
+                # all 3 parameters are callable
+                if sum(is_callables) == 3:
+                    reordered_params.append(
+                        [params[params_idx], params[params_idx + 1], params[params_idx + 2]]
+                    )
+                    reordered_params.append(
+                        [params[params_idx], params[params_idx + 1], params[params_idx + 2]]
+                    )
+                    coeff_idx += 2
+                    params_idx += 3
+
+                # 2 of 3 parameters are callable
+                elif sum(is_callables) == 2:
+                    reordered_params.append([params[params_idx], params[params_idx + 1]])
+                    reordered_params.append([params[params_idx], params[params_idx + 1]])
+                    coeff_idx += 2
+                    params_idx += 2
+
+                # 1 of 3 parameters is callable
+                elif sum(is_callables) == 1:
+                    reordered_params.append(params[params_idx])
+                    reordered_params.append(params[params_idx])
+                    coeff_idx += 2
+                    params_idx += 1
+
+                # in case of no callable, the coeff is still callable due to the explicit freq*t dependence
+                elif sum(is_callables) == 0:
+                    reordered_params.append(None)
+                    reordered_params.append(None)
+                    coeff_idx += 2
+                    params_idx += 0
```

Comment:
I just noticed this: :innocent: 
```suggestion
                # all 3 parameters are callable
                num_callables = sum(is_callables)
                reordered_params.extend([params[params_idx : params_idx + num_callables]]*2)
                coeff_idx += 2
                params_idx += num_callables
```

Reply:
(Could also keep the two lines that do `reordered_params.append` instead of switching to `extend`, of course :) )

Reply:
🧠 🤯

Reply:
needed a little patch above in the code (`params[0]` in the case of one single parameter)
now implemented and tests passing
amazing upgrade!

Reply:
Ah, do you mean the case of `num_callables=0`? :thinking: 
Great :)

Reply:
No, for the case of one single parameter. The reorder function would turn `[1] |-> [1, 1]` and have the callables just take the param from there, but this altered version always created lists of params, so in that case would do `[1] |-> [[1], [1]]`, so needed to add a `params[0]` to take the first element, but that is more consistent overall anyway in the way the function works, so improvement overall 👍

Reply:
Ah, now I see :) Thanks! In the case of 0 parameters we're now using empty lists instead of `None`, right? That should be fine, I suppose? :)